Macros/Macros.js

```diff
@@ -192,6 +223,64 @@ class FurnaceMacros {
         div.find(".furnace-macro-expand").on('click', (ev) => div.toggleClass("fullscreen"));
         refreshHighlight();
     }
+
+    static async _execute(message) {
+        let macro = game.macros.getName(message.macroName);
+        if (!macro || !getProperty(macro.data, "flags.furnace.gmexecute" || !game.users.get(macro.data.author).isGM)) {
+            ui.notification.error(`User ${message.userId} tried to invailidly GM execute macro ${message.macroName}`);
+            return "Invalid Request";
+        }
+
+        try {
+            return await macro.renderContent("RemoteContext", message, ...message.args);
+        } catch (err) {
+            ui.notifications.error(`There was an error in your macro syntax. See the console (F12) for details`);
+            console.error(err);
+            return "Macro syntax error";
+        }
+    }
+
+    // request remote execution of a macro
+    static async requestRemoteExecute(macroName, ...args) {
+        let targetGM = game.users.entities.find(u => u.isGM && u.active);
+        let macroMessage = {
+            action: "GMExecuteMacro", 
+            requestId: "remoteResult" + game.user.id + Date.now(), 
+            userId: game.user.id, 
+            targetGM: targetGM._id, 
+            macroName, 
+            sceneId: canvas.scene.id, // pass this so macros can know which scene the player is on.
+            tokenId: ChatMessage.getSpeaker().token, 
+            event: event || {event: {shiftKey: true}}, // this is to keep minor-qol happy
+            args
+        };
+        if (game.user.isGM) return await FurnaceMacros._execute(macroMessage); 
```

Comment:
We already checked that we're not GM before calling this function. I'd rather check here that the macro is a GM macro and has all the right permissions, because then we would run it locally instead of sending it to the GM who will then answer with a permission error.